### PR TITLE
[Bug] Allow exact prefix matches for Mamba radix cache #21158

### LINKED ISSUE #21158: [Bug] The cache hit rate of Qwen3.5 MambaRadixTree remains zero when i use  the same prompt.
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When I repeatedly use the same prompt, the cache hit rate of Qwen3.5 MambaRadixTree remains zero. Is this expected behavior?
It seems this is caused by the logic in schedule_batch.py, where max_prefix_len = input_len - 1 and token_ids = self.fill_ids[:max_prefix_len].
This results in fewer matched nodes during the matching phase, which leads to node splitting and ultimately causes the cache hit count to remain zero.
Is this consistent with the intended design?

### Reproduction

python3 -m sglang.launch_server --model-path /data/models/Qwen3___5-35B-A3B-FP8/ --host 0.0.0.0 --tp 2   --enable-metrics --port 8001 --enable-cache-report --cuda-graph-max-bs 64  --page-size 1

### Environment

root@lts-4090:/code/sglang-v0.5.9# python3 -m sglang.check_env
Python: 3.12.11 (main, Jun  4 2025, 08:56:18) [GCC 11.4.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA GeForce RTX 4090
GPU 0,1,2,3,4,5,6,7 Compute Capability: 8.9
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.6, V12.6.68
CUDA Driver Version: 565.57.01
PyTorch: 2.9.1+cu128
sglang: 0.0.0.dev10696+gd20e9a20f.d20260319
sglang-kernel: 0.4.0
flashinfer_python: 0.6.6
flashinfer_cubin: 0.6.6
flashinfer_jit_cache: Module Not Found
triton: 3.5.1
transformers: 4.57.1
torchao: 0.9.0+cu126
numpy: 2.3.2
aiohttp: 3.12.15
fastapi: 0.116.1
hf_transfer: 0.1.9
huggingface_hub: 0.34.4
interegular: 0.3.3
modelscope: 1.29.1
orjson: 3.11.2
outlines: 0.1.11
packaging: 25.0
psutil: 7.0.0
pydantic: 2.11.7
python-multipart: 0.0.20
pyzmq: 27.0.2
uvicorn: 0.35.0
uvloop: 0.21.0
vllm: Module Not Found
xgrammar: 0.1.27
openai: 2.6.1
tiktoken: 0.11.0
anthropic: 0.64.0
litellm: Module Not Found
torchcodec: 0.9.1
NVIDIA Topology:
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	PHB	PHB	PHB	SYS	SYS	SYS	SYS	0-57	0		N/A
GPU1	PHB	 X 	PHB	PHB	SYS	SYS	SYS	SYS	0-57	0		N/A
GPU2	PHB	PHB	 X 	PHB	SYS	SYS	SYS	SYS	0-57	0		N/A
GPU3	PHB	PHB	PHB	 X 	SYS	SYS	SYS	SYS	0-57	0		N/A
GPU4	SYS	SYS	SYS	SYS	 X 	PHB	PHB	PHB	58-115	1		N/A
GPU5	SYS	SYS	SYS	SYS	PHB	 X 	PHB	PHB	58-115	1		N/A
GPU6	SYS	SYS	SYS	SYS	PHB	PHB	 X 	PHB	58-115	1		N/A
GPU7	SYS	SYS	SYS	SYS	PHB	PHB	PHB	 X 	58-115	1		N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

Hypervisor vendor:: KVM
ulimit soft: 1048576

### PR DESCRIPTION
**Pull request by robellliu-dev**
## Summary
- allow Mamba radix cache requests to match the full prompt when prompt logprobs are not requested
- keep the existing one-token-short behavior for non-Mamba caches and logprob requests
- add a regression test covering repeated prompts with `Req.init_next_round_input()` and `MambaRadixCache`

## Root Cause
`Req.init_next_round_input()` always capped prefix matching at `input_len - 1`. That behavior is needed for non-Mamba caches so the scheduler can recompute the last prompt token for logprob / first-token generation. But `MambaRadixCache` stores reusable recurrent state on whole-sequence nodes and cannot split the last token into a reusable Mamba state. Repeated prompts therefore only matched a shortened prefix, split the leaf into an internal node with no `mamba_value`, and never reached a reusable Mamba cache hit.

## Fix
When the tree cache supports Mamba and prompt logprobs are not requested, allow exact prefix matches by setting the matchable prefix length to the full input length. This preserves the old `input_len - 1` behavior for non-Mamba caches and for logprob requests.

## Testing
- `python -m py_compile python/sglang/srt/managers/schedule_batch.py test/registered/unit/mem_cache/test_mamba_unittest.py`
- `python -m unittest test/registered/unit/mem_cache/test_mamba_unittest.py` *(blocked locally on Windows because this repo imports Unix-only `resource` during module initialization)*

Fixes #21158

### CODE REVIEW

File: test/registered/unit/mem_cache/test_mamba_unittest.py [hunk 1]
```diff
@@ -457,6 +457,38 @@ def test_insert_prev_prefix_len(self):
...
+        tree.cache_finished_req(req1)
+
+        req2 = make_req("req2")
+        req2.init_next_round_input(tree)
+        assert len(req2.prefix_indices) == len(req2.origin_input_ids)
+        assert req2.last_node.mamba_value is not None
+
+        req3 = make_req("req3", return_logprob=True)
+        req3.init_next_round_input(tree)
+        assert len(req3.prefix_indices) == len(req3.origin_input_ids) - 1
```
[the remark this answers is not in the document]
Comment (robellliu-dev) [reply test/registered/unit/mem_cache/test_mamba_unittest.py]: Good catch. That assertion was incorrect.

`req3` sets `return_logprob=True`, and with the default `logprob_start_len=0` the matched prefix length collapses to 0, so the test expectation was wrong. After tracing the broader behavior, I ended up backing out the full-prefix Mamba change from this branch entirely rather than trying to preserve this test.

File: python/sglang/srt/managers/schedule_batch.py [hunk 2]
```diff
@@ -873,10 +873,17 @@ def init_next_round_input(
...
+        # prompt logprobs / first-token logits from the last prompt token. Mamba caches can
+        # reuse the full recurrent state, so allow an exact prefix hit when logprobs are not
+        # requested.
+        supports_mamba = tree_cache is not None and tree_cache.supports_mamba()
+        max_prefix_len = input_len if supports_mamba else input_len - 1
         if self.return_logprob and self.logprob_start_len >= 0:
             max_prefix_len = min(max_prefix_len, self.logprob_start_len)
+        elif not supports_mamba:
+            # Keep at least one uncached token for non-Mamba models.
+            max_prefix_len = min(max_prefix_len, input_len - 1)
```
[the remark this answers is not in the document]
Comment (robellliu-dev) [reply python/sglang/srt/managers/schedule_batch.py]: Agreed. After tracing this through the current scheduler path, the exact full-prefix Mamba behavior is not safe in this PR.

Even with a simplified `max_prefix_len` expression, allowing `max_prefix_len == input_len` can still produce a zero-token extend, which the current EXTEND path does not handle correctly. I backed out the exact full-prefix change from this branch instead of trying to refine that logic here.

File: python/sglang/srt/managers/schedule_batch.py [hunk 3]
```diff
@@ -873,10 +873,17 @@ def init_next_round_input(
...
 
         input_len = len(self.fill_ids)
-        # NOTE: the matched length is at most 1 less than the input length to enable logprob computation
-        max_prefix_len = input_len - 1
+        # NOTE: Non-Mamba caches leave one token uncached so the next prefill can compute
+        # prompt logprobs / first-token logits from the last prompt token. Mamba caches can
+        # reuse the full recurrent state, so allow an exact prefix hit when logprobs are not
+        # requested.
+        supports_mamba = tree_cache is not None and tree_cache.supports_mamba()
+        max_prefix_len = input_len if supports_mamba else input_len - 1
```
[the remark this answers is not in the document]
Comment (robellliu-dev) [reply python/sglang/srt/managers/schedule_batch.py]: Thanks, I agree this is the blocking issue.

Allowing an exact full Mamba cache hit here can produce `extend_input_len == 0`, and the current EXTEND / logits path assumes every extend request has at least one token. Rather than trying to thread zero-token extend support through this PR, I backed out the exact full-prefix Mamba change from this branch. The PR now has no net diff against main, and I think the right follow-up is a separate design for safe zero-token-extend handling (or another safe way to reuse Mamba state).

**Comment by hnyls2002** [comment]
Hi @robellliu-dev, this PR currently shows no file changes (+0/-0 diff), so there is nothing for it to merge even though the description outlines a real fix. Could you re-push the intended change to Req.init_next_round_input() and MambaRadixCache? I will close this empty PR for now -- please open a fresh one with the actual diff.
